### ISSUE TITLE
Add Concept Maps

### ISSUE THREAD

**Issue by anewage** [InsightfulSummer/alphaLib-web#27]
In order to visualize the network of ideas and concepts (in our case, words), we need a simple network diagram component. Input to this component includes both nodes and links. You can view more about this in `components/ConceptMap`. Input to this component should be similar to the following:
```javascript
  data: {
    nodes: [
      { name: 'idea1', from:'query', size: 10 },
      { name: 'idea2', from:'document', docID: '<some-id>', size: '20' },
      ...
    ],
    links: [
      { source: 'idea1', target: 'idea2', value: 7 },
      ...
    ]
  }
```

The above is only a sample data and it is up to you to decide on the final structure of the data.

Apart from concept map, this visualization would be used to visualize any data described in nodes and links. In developing the core component (`TheGraph.vue`) please add labels to the middle of the circles.

Note: **Please keep in mind that the criteria for having a "relation" or a "link" is separate from the logic of displaying  a given set of nodes and links.** For now, we are saying that if a word is in a sentence where another word also appears, we have a relation. But this process is being handled in the wrapper component and is separate from the core component.

A general rule of thumb and a guideline for implementation: use "driver" design pattern. This design pattern separates the idea of the actual process and functionality from its environment (the types of data, different kinds of instantiation, ...).
To be precise, it is highly recommended to have:
- a stand-alone "graph" component inside `components`. This component only deals with primitive SVG tags and D3.js code is actually happening inside it. It has a specific set of props (data inputs) through which it can be controlled from outside. Some data similar to the input described above would be passed through props to this component and it would display the graph.
- a "wrapper" component which is in charge of data transformations and visual fitting of that the actual graph component. One very obvious feature of this wrapper is to support the specific data structure that the actual bar chart requires and transform the data which is coming from the page element to that specific format (nodes and links). Another is to scaffold visual components to support the styling and visual representation of this element (In the case of current components, the wrapper component is a `v-card` and hugs the actual component inside it!)
(As an example, please take a look at Graph or Heatmap components currently in `dev` branch).

A basic breakdown of this feature could be:
- [x] displaying a list of nodes and links (the above input) as a generic, simple graph
- [ ] concertizing the data structure and the code into one coherent component (`TheBarChart.vue`)
- [ ] creating a wrapper component and scaffolding the data and visual transformations + connecting it to data coming from page components
- [ ] finalizing the components and instantiating the wrapper component in the proper place on page components.

Note: D3.js and Vue.js differ in their general approach to DOM management. Please make sure to check the code for `TheHeatMap.vue` component to understand this difference and please try to keep your code as much aligned with this style (style of `TheHeatMap.vue` -- mentioning svg tags in template as much possible and avoiding pure D3.js scripts) as possible.

Some examples from D3's gallery (https://observablehq.com/@d3/gallery):
- https://observablehq.com/@d3/force-directed-graph
- https://observablehq.com/@bstaats/graph-visualization-introduction
- https://observablehq.com/@d3/disjoint-force-directed-graph
- https://observablehq.com/@fil/graph-voronoi
- https://observablehq.com/@xianwu/simple-force-directed-graph-network-graph
- https://observablehq.com/@mbostock/nearest-neighbor-graph
- https://observablehq.com/@vk2425/game-of-thrones-relationship-graph
- https://observablehq.com/@stwind/github-files-graph
- https://observablehq.com/@nhogs/zdog-3d-force-directed-graph
- https://observablehq.com/@turuibo/the-complete-causal-graph-of-neuropathic-pain-diagnosis

Note: The development of this component is separate from the rule of "having a relationship". Different rules would act as differnet applications and instantiations of graph component on different pages and locations. **Having reusable and flexible components is the most important focus of this project.**

**Comment by DogsAreTheShit** [comment]
**Question: How do all the files interact with each other**

**What I understand**
From the issue you wrote, what I gathered was that ‘theGrath.vue’ file takes formatted data (ie: nodes and links) from props and creates SVG’s / graphs with that data. Also from my understanding, ‘GraphWrapper.vue’ takes the raw documents and processes them (amongst other tasks), turning them into properly formatted data (nodes and links) to be passed into ‘TheGraph.vue’ props.

**Where I need a little more clarification is:**
How exactly do props work?
Which file is the ‘prop’ functionality implemented? I see props in ‘TheGraph.vue’ file and ‘GraphWrapper.vue’ file
How do the other components fall into this? 

I think the best way I will understand is through a flow chart that explains the hierarchy of all these components, their input/outputs, etc… but if there is a better/easier way you know of that works too

**Comment by DogsAreTheShit** [comment]
@anewage

**Comment by DogsAreTheShit** [comment]
Another question, for the second point on the checklist, do you want me to create the component TheBarChart.vue? or is that just an example

**Comment by DogsAreTheShit** [comment]
Currently trying to get labels to display beside nodes. I just committed my current progress, trying some changes after research on adding labels to nodes, not working currently. If you have time to check out my commit and see the problem, i'd appreciate it

**Comment by anewage** [comment]
> 
> 
> **Question: How do all the files interact with each other**
> 
> **What I understand**
> From the issue you wrote, what I gathered was that ‘theGrath.vue’ file takes formatted data (ie: nodes and links) from props and creates SVG’s / graphs with that data. Also from my understanding, ‘GraphWrapper.vue’ takes the raw documents and processes them (amongst other tasks), turning them into properly formatted data (nodes and links) to be passed into ‘TheGraph.vue’ props.
> 
> **Where I need a little more clarification is:**
> How exactly do props work?
> Which file is the ‘prop’ functionality implemented? I see props in ‘TheGraph.vue’ file and ‘GraphWrapper.vue’ file
> How do the other components fall into this?
> 
> I think the best way I will understand is through a flow chart that explains the hierarchy of all these components, their input/outputs, etc… but if there is a better/easier way you know of that works too

Almost every component that we have is now controlled through props. So consider the following example:
```vue
// somePage.vue
<template>
  <some-component :prop1="something" :prop2="someOtherThing" />
</template>
```
and then inside the component itself:
```vue
// SomeComponent.vue
<template>
  ...
  <div :class="something">{{ someOtherThing }}</div>
  ...
</template>
<script>
  export default {
    ...
    props: {
      something: {
        type: String,
        default: ""
      },
      someOtherThing: {
        type: String,
        default: ""        
      }
    },
    ...
  }
</script>
```
This creates a one-way data stream from the parent component (in this case the page component) to the child component (the component being instantiated inside `template` of a parent). More about props: https://vuejs.org/v2/guide/components-props.html

**Comment by DogsAreTheShit** [comment]
@anewage So from my understanding, currenlty _GraphWrapper.vue_ has sample data already in the form of nodes and links and is passing this sample data to _TheGraph.vue_ to be turned into the visual diagram. The two areas that need work are:
**_TheGraph.vue_**: needs labels inside of nodes
_**GraphWrapper.vue**_: replace sample data with docs, which then need to be transformed into nodes and links

Of those two, I'm going to prioritize replacing sample data
**Two Questions: **
1. first im just worrying about splicing the document name/id/etc info, not the actual contents of the documents yet correct?
2. What is the best way for my purposes to access the data? I'm guessing some code for basic splicing of the data is already written for the results display page(ie:  this is the title, this is the author etc)

**Comment by DogsAreTheShit** [comment]
@aanewage
1.  I'm looking at the raw data on the documents and (going off of my java experience) it appears to already be in an array/dictionary format? (ie: {"title":"design for a brain", "score": "1", etc...})
Is there a way I can access this already in the form of an array? or do I have to treat it as a long string and splice it at the commas. I see in pages/docs/index.vue it accesses this info through {{.docs._source.meta.title }} but I can't use that in the script section

2. Question about how you want me to format the data into nodes and links, should it be structured like
      i) Nodes are documents and links connect documents based off of shared concepts
      or
      ii) Nodes are concepts and the links that connect the nodes is based on how often these concepts appear together

**Comment by DogsAreTheShit** [comment]
Also what functionality is currently still missing for _thegraph.vue_ file. I see that it properly visualizes the nodes and links and that it still needs labels implemented. Other than that is _thegraph.vue_ still missing other features or is the rest of the work in the _graphwrapper.vue_

**Comment by DogsAreTheShit** [comment]
@anewage Hey Amir, gentle reminder if you have the time to answer any of my questions

**Comment by DogsAreTheShit** [comment]
@anewage 

Can you take a look at my last couple of commits and give me some pointers. I'm trying to get text to appear beside the nodes but it isnt displaying

**Comment by DogsAreTheShit** [comment]
Also letting you know my current plan to accomplish my task, let me know if you have any comments
1. transform contents of docs into word arrays -
2. each word has a 'link' to each other word that it appeared with within the same sentence
3. Links are stronger the more two words appear together
4. only links of x strength will be displayed on the concept map

**Comment by DogsAreTheShit** [comment]
I've made progress on the issue but am still gated by not knowing how to declare nodes and links. If you could fill me in on that it would be much appreciated. Let me know if you need more explanation on what Im asking

**Comment by DogsAreTheShit** [comment]
@anewage Ok ive figured out the other issues, I just have the one question now

I tried using this.doc._source.meta.title  and I got this error: 
<img width="720" alt="Screen Shot 2020-07-14 at 1 24 12 PM" src="https://user-images.githubusercontent.com/65428292/87456879-88ad8300-c5d5-11ea-80c9-9712958f2c11.png">
<img width="432" alt="Screen Shot 2020-07-14 at 1 28 09 PM" src="https://user-images.githubusercontent.com/65428292/87457104-dd50fe00-c5d5-11ea-8736-8cdfbfd7297c.png">
let me know if you know what I'm doing wrong. ill commit the changes so you can see how I am using the code as well with the screen shots above

**Comment by DogsAreTheShit** [comment]
I used the line `console.log('dataset[1] _id value >>>>', this.dataset[1]._id)` and it properly printed to console:
<img width="465" alt="Screen Shot 2020-07-17 at 5 32 37 PM" src="https://user-images.githubusercontent.com/65428292/87832221-82c4d580-c853-11ea-9070-e1c950ecf292.png">

but for some reason it makes the page stop working:
![Screen Shot 2020-07-17 at 5 32 55 PM](https://user-images.githubusercontent.com/65428292/87832243-907a5b00-c853-11ea-8e91-35d63b73b582.png)

**Comment by anewage** [comment]
So whenever you put a static reference to something in the code, specially when that reference is to an array element, you need to consider scenarios when that array is empty. When the component is being rendered, the data is not yet processed and hence there is no item in the 1st index of `this.dataset`.

If you had a static reference only to an empty object, it would show `undefined` without any errors. But since you are also trying to access a property of that empty element, namely the `_id` attribute, you see an error.

In order to fix this, you need to process arrays with for loops.

**Comment by DogsAreTheShit** [comment]
Awesome working now, another issue:
How do I pass all the nodes I've created to the main component? 
Currently, I'm passing a specified element of my node array and it correctly is displayed by the main component:

<img width="766" alt="Screen Shot 2020-07-17 at 7 04 46 PM" src="https://user-images.githubusercontent.com/65428292/87837094-62e7de80-c860-11ea-86a1-4ec73d7e904d.png">

 but how do I pass the whole array? Tried this but it didn't work
<img width="226" alt="Screen Shot 2020-07-17 at 7 05 03 PM" src="https://user-images.githubusercontent.com/65428292/87837111-6c714680-c860-11ea-9283-c9290803acdd.png">

**Comment by DogsAreTheShit** [comment]
Also, how do I access the contents of a document? Currently, my test object has a paragraph of text which is passed on, then the nouns are isolated and used as the 'concepts' of a given document to be linked as nodes. However in changing over from my test object to the real data, I don't see a data value for the contents of a document (like _id etc):
![Screen Shot 2020-07-17 at 7 11 19 PM](https://user-images.githubusercontent.com/65428292/87837374-4e581600-c861-11ea-85b4-51c59242f58b.png)

**Comment by DogsAreTheShit** [comment]
Hm the issue is back. I put the id request inside a for loop but I'm getting the same issue
**Here is the for loop: **
<img width="447" alt="Screen Shot 2020-07-17 at 9 13 33 PM" src="https://user-images.githubusercontent.com/65428292/87841294-5ff5e980-c872-11ea-9c4a-1a4777f439b6.png">
**Here is the error: **
<img width="594" alt="Screen Shot 2020-07-17 at 9 14 08 PM" src="https://user-images.githubusercontent.com/65428292/87841306-74d27d00-c872-11ea-95eb-22611a641bf9.png">

**Comment by DogsAreTheShit** [comment]
@anewage

**Comment by DogsAreTheShit** [comment]
Okay so almost everything is working as intended

Current build:
because of the issue I'm having where the code executes before the dataset is defined, I wrote it so that if the dataset is undefined, it uses sample data, and once the dataset is defined, it uses the dataset. So what happens is when I load the page, my wrapper is executed once, uses the sample data, builds the nodes, passes them to the main component, then my wrapper is executed AGAIN, uses the real data now (because at this point the dataset has been defined), builds nodes and links off of dataset, but doesn't pass the new nodes and links to the main component.

I can better explain this issue If you have the time for a zoom call. Let me know or ill wait til our next zoom call

**Comment by DogsAreTheShit** [comment]
@anewage currently to get my graph to load I'm using a watch component, so when the links change, the graph is drawn and it works but when you do multiple searches, the 'outdated' graph doesn't disappear, leading to both graphs overlaying. is there a way for me to clear the old nodes and links before I redraw the new ones in my watch component?
